### DISCUSSION TITLE
IndigoRender rotate image

### DISCUSSION THREAD

**Discussion by twall** [epam/Indigo#1906] [Q&A] [answered]
Are there any options to force a rotation of a rendered molecule?  I'd like the option to request rendering of an arbitrary rotation of the structure, leaving the labels in the correct orientation.  If I'm looking at two molecules side by side I'd like to be able to orient them in a way that highlights similarity.

Ketcher allows rotation of selected molecules.

<img width="156" alt="Screenshot 2024-04-08 at 5 31 40 PM" src="https://github.com/epam/Indigo/assets/756909/cfb35abc-8ba5-4c00-99c4-d88daf14ddab">
<img width="149" alt="Screenshot 2024-04-08 at 5 31 55 PM" src="https://github.com/epam/Indigo/assets/756909/6729ae2c-a4bd-4f7c-9737-f7f40f2a2492">

**Comment by AlexanderSavelyev** [comment]
Hi @twall . It is not clear what exactly is required. It seems like not rendering but re-layout a molecule with appropriate angle and center of the rotation. Is it correct? If you want to highlight substructure, one can use highlightedTarget
```
match = indigo.substructureMatcher(mol).match(query)

if match:
   match.highlightedTarget().saveMolfile("highlighted.mol")
```

**Comment by twall** [comment]
Suppose I’m displaying two identical molecules but for whatever reason one
is rendered upside down.  How do I correct that?  I can do a CSS transform,
but that leaves the labels inverted.

Ideally I would be able to automatically align and highlight similarities
of two molecules, but I don’t think that’s possible currently without some
extra input.  So my starting point is to just enable the user to rotate one
or the other images.

**Comment by twall** [comment]
Thank you for the highlighting tip, though, I was not aware of that
functionality (the docs, particularly for nosql, are rather thin).

**Comment by AlexanderSavelyev** [comment, marked as answer]
As a solution we can suggest to use XYZ function for atoms and rotate using standard rotation matrix. See the following code as an example

```
def rotate(x, y, angle):
    xn = x * math.cos(angle) - y * math.sin(angle)
    yn = x * math.sin(angle) + y * math.cos(angle)
    return xn, yn


m = indigo.loadMolecule('C1([H])C=CC=C1')
m.layout()

indigoRenderer.renderToFile(m, 'result1.png')

angle = 3.14159
for a in m.iterateAtoms():
    xyz = a.xyz()
    xn, yn = rotate(xyz[0], xyz[1], angle)
    a.setXYZ(xn, yn, 0)

indigoRenderer.renderToFile(m, 'result2.png')
```

![result1](https://github.com/epam/Indigo/assets/7427808/aafcb5ee-0f64-4cbe-b9d8-07e2ac7094a8)
![result2](https://github.com/epam/Indigo/assets/7427808/07dad159-2ff8-4f7b-a850-c7f23c7bf8a0)
